### PR TITLE
Added version support to Razor parser

### PR DESCRIPTION
**Pull request by ajaybhargavb**
#1678 

@rynowak 

- Added `Version` property to `RazorParserOptions`
- Added an internal `RazorParserFeatureContext` that contains a flag for all features we want to enable/disable in the parser
- Reason for having a centralized location is so we don't have to look at the version every time we want to decide whether to light-up a feature
- ~~I'll send a separate PR that will use `AllowMinimizedBooleanTagHelperAttributes` to enable/disable that feature.~~ https://github.com/aspnet/Razor/pull/1695

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Language/DefaultRazorParserOptions.cs [hunk 1]
```diff
@@ -18,12 +18,18 @@ public DefaultRazorParserOptions(DirectiveDescriptor[] directives, bool designTi
             Directives = directives;
             DesignTime = designTime;
             ParseLeadingDirectives = parseLeadingDirectives;
+            Version = version ?? RazorParserFeatureContext.LatestRazorVersion;
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/DefaultRazorParserOptions.cs]: Should this throw ArgNull instead?

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureContext.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    internal abstract class RazorParserFeatureContext
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureContext.cs]: Couldn't think of a better that isn't already used

File: src/Microsoft.AspNetCore.Razor.Language/DefaultRazorParserOptions.cs [hunk 3]
```diff
@@ -18,12 +18,18 @@ public DefaultRazorParserOptions(DirectiveDescriptor[] directives, bool designTi
...
 
         public override bool DesignTime { get; }
 
         public override IReadOnlyCollection<DirectiveDescriptor> Directives { get; }
 
         public override bool ParseLeadingDirectives { get; }
+
+        public override Version Version { get; }
+
+        internal override RazorParserFeatureContext FeatureContext { get; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/DefaultRazorParserOptions.cs]: I would drop the name `Context` from this. It's not really acting as a context parameter. 

So maybe: `internal override RazorParserFeatureFlags FeatureFlags { get; set; }`

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureContext.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        internal static readonly Version LatestRazorVersion = new Version(2, 1, 0);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureContext.cs]: Think about using an enum here instead. Version allows you to express a huge range of stuff including versions that done exist.

Think continuous vs discrete.

We never need/want to handle a version we didn't ship.

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserOptions.cs [hunk 5]
```diff
@@ -54,5 +54,9 @@ public static RazorParserOptions CreateDesignTime(Action<RazorParserOptionsBuild
         /// In a future release this may be updated to include all leading directive content.
         /// </remarks>
         public abstract bool ParseLeadingDirectives { get; }
+
+        public abstract Version Version { get; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorParserOptions.cs]: this is a breaking change

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserOptions.cs [hunk 6]
```diff
@@ -54,5 +54,9 @@ public static RazorParserOptions CreateDesignTime(Action<RazorParserOptionsBuild
...
+
+        internal abstract RazorParserFeatureContext FeatureContext { get; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorParserOptions.cs]: this is a breaking change

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserOptionsBuilder.cs [hunk 7]
```diff
@@ -13,6 +14,8 @@ public abstract class RazorParserOptionsBuilder
 
         public abstract bool ParseLeadingDirectives { get; set; }
 
+        public abstract Version Version { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorParserOptionsBuilder.cs]: breaking change,

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserOptions.cs (hunk 5 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Razor.Language/RazorParserOptions.cs]: These need to be virtual

**Comment by ajaybhargavb** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserVersion.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    public enum RazorParserVersion
+    {
+        Version1_0,
+        Version1_1,
+        Version2_0,
+        Version2_1,
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/RazorParserVersion.cs]: Will there be a need to add more here other than for every major release? Eg. We'll need different parser behavior for patch releases?

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureFlags.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    internal abstract class RazorParserFeatureFlags
+    {
+        internal static readonly RazorParserVersion LatestRazorParserVersion = RazorParserVersion.Version2_1;
+
+        public static RazorParserFeatureFlags Create(RazorParserVersion version)
+        {
+            if (version == LatestRazorParserVersion)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureFlags.cs]: This should be a >= `Version2_1` - we do don't want to keep changing this code if we add more versions

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureFlags.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    internal abstract class RazorParserFeatureFlags
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureFlags.cs]: This doesn't really need to be abstract if its internal

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserVersion.cs (hunk 8 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Razor.Language/RazorParserVersion.cs]: No, not for patches

Comment (rynowak) [reply src/Microsoft.AspNetCore.Razor.Language/RazorParserVersion.cs]: We might want to do these the way Roslyn does - https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/LanguageVersion.cs

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserVersion.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    public enum RazorParserVersion
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorParserVersion.cs]: RazorLanguageVersion

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureFlags.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        internal static readonly RazorParserVersion LatestRazorParserVersion = RazorParserVersion.Version2_1;
+
+        public static RazorParserFeatureFlags Create(RazorParserVersion version)
+        {
+            if (version == LatestRazorParserVersion)
+            {
+                return new DefaultRazorParserFeatureFlags(allowMinimizedBooleanTagHelperAttributes: true);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureFlags.cs]: Imagine you have 4-5 different features to control. Think about just having them as variables and then call the constructor at the end.

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureFlags.cs (hunk 10 above)
Comment (ajaybhargavb) [reply src/Microsoft.AspNetCore.Razor.Language/RazorParserFeatureFlags.cs]: I didn't want this to be constructible without using the Create method.

File: src/Microsoft.AspNetCore.Razor.Language/DefaultRazorParserOptionsBuilder.cs [hunk 13]
```diff
@@ -19,9 +20,11 @@ public DefaultRazorParserOptionsBuilder(bool designTime)
 
         public override bool ParseLeadingDirectives { get; set; }
 
+        public override RazorParserVersion Version { get; set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/DefaultRazorParserOptionsBuilder.cs]: We should treat `Version` like we do `DesignTime`, aka required at construction of the builder.

**Comment by ajaybhargavb** [comment]
🆙 📅

**Pull request merged**